src/Haskell/Ide/Engine/Plugin/Ormolu.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE CPP #-}
+
+module Haskell.Ide.Engine.Plugin.Ormolu ( ormoluDescriptor ) where
+
+import Data.Aeson ( Value ( Null ) )
+import Data.Text
+import Haskell.Ide.Engine.MonadTypes
+
+#if __GLASGOW_HASKELL__ >= 806
+import Control.Exception
+import Control.Monad.IO.Class ( liftIO , MonadIO(..) )
+import Ormolu
+import Ormolu.Config (defaultConfig)
+import Ormolu.Exception (OrmoluException)
+import Haskell.Ide.Engine.PluginUtils
+#else
+import Haskell.Ide.Engine.MonadFunctions
+#endif
+
+ormoluDescriptor :: PluginId -> PluginDescriptor
+ormoluDescriptor plId = PluginDescriptor
+  { pluginId                 = plId
+  , pluginName               = "Ormolu"
+  , pluginDesc               = "A formatter for Haskell source code."
+  , pluginCommands           = []
+  , pluginCodeActionProvider = Nothing
+  , pluginDiagnosticProvider = Nothing
+  , pluginHoverProvider      = Nothing
+  , pluginSymbolProvider     = Nothing
+  , pluginFormattingProvider = Just provider
+  }
+
+#if __GLASGOW_HASKELL__ >= 806
+provider :: FormattingProvider
+provider contents uri typ _opts =
+  case typ of 
+    FormatRange _ -> return $ IdeResultFail (IdeError PluginError (pack "Selection formatting for Ormolu is not currently supported.") Null)
+    FormatText -> pluginGetFile contents uri $ \file -> do
+        result <- liftIO $ try @OrmoluException (ormolu defaultConfig file (unpack contents))
+        case result of
+          Left  err -> return $ IdeResultFail (IdeError PluginError (pack $  "ormoluCmd: " ++ show err) Null)
+          Right new -> return $ IdeResultOk [TextEdit (fullRange contents) new]
+#else
+-- Work in progress
+provider :: FormattingProvider
+provider _contents _uri typ _opts = do
+  errorm "This version of HIE does not support Ormolu as a formatter"
+  case typ of 
+    FormatRange _ -> return $ IdeResultFail (IdeError PluginError (pack "Selection formatting for Ormolu is not currently supported.") Null)
```

Comment:
This should be like FormatText a noop if it is not supported.

Reply:
What about notifying about lack of support for GHC < 8.6? (The use of `errorm` included in this snippet.)

Reply:
That should be done somewhere else imo, I am not sure where, yet, but not here. @bubba would be server.hs fine?